cpp/src/arrow/compute/kernels/codegen_internal.h

```diff
@@ -322,6 +329,26 @@ struct ArrayIterator<Type, enable_if_base_binary<Type>> {
   }
 };
 
+template <typename Type>
+struct ArrayIterator<Type, enable_if_list_type<Type>> {
+  using T = typename TypeTraits<Type>::ScalarType;
+  using ArrayT = typename TypeTraits<Type>::ArrayType;
+  using offset_type = typename Type::offset_type;
+
+  const ArraySpan& arr;
+  int64_t position;
+
+  explicit ArrayIterator(const ArraySpan& arr) : arr(arr), position(0) {}
+
+  T operator()() {
+    const auto array_ptr = arr.ToArray();
```

Comment:
The alternative to calling `ToArray` with the cast would be to implement something like `value_slice` on the `ArraySpan` directly, although I'm not sure if the `ArraySpan` is supposed to return anything but pointers to primitives (as is currently implemeted)

Reply:
This is going to be slow, so we probably want to avoid this IMHO.

You may want to run a crude benchmark from Python to check this.

Reply:
How about:
1. Get offset, length
2. Subslice the value array
3. Build `ListScalar` / `LargeListScalar` from the child array?

Or materialize the child array, and using the sub array. `arr.ToArray()` every call is too expansive?

Reply:
Sorry for the late reply - I somehow missed this comment. So the expectationn is that the ArraySpan -> Array conversion is expensive here and getting a slice first before making that conversion should help with performance, right?

Reply:
The conversion will be expensive regardless of whether you slice before or afterwards, so ideally we should avoid the conversion entirely.

Reply:
In other words, we should consider exposing a `ArrayRangeEquals` variant that takes `ArraySpan` inputs instead of `Array` (the underlying implementation can be converted to `ArraySpan` entirely).